worker/rhpv2.go

```diff
@@ -262,6 +266,225 @@ func (w *worker) FetchSignedRevision(ctx context.Context, hostIP string, hostKey
 	return rev, err
 }
 
+func (w *worker) DeleteContractRoots(ctx context.Context, hostIP, hostVersion string, hostKey types.PublicKey, renterKey types.PrivateKey, contractID types.FileContractID, lastKnownRevisionNumber uint64, timeout time.Duration, indices []uint64) (int64, error) {
+	w.logger.Debugw(fmt.Sprintf("deleting %d contract roots (%v)", len(indices), humanReadableSize(len(indices)*rhpv2.SectorSize)), "hk", hostKey, "fcid", contractID)
+
+	// escape early if no indices are given
+	if len(indices) == 0 {
+		return 0, nil
+	}
+
+	// sort in descending order so that we can use 'range'
+	sort.Slice(indices, func(i, j int) bool {
+		return indices[i] > indices[j]
+	})
+
+	// decide on the batch size, defaults to ~20mib of sector data but for old
+	// hosts we use a much smaller batch size to ensure we nibble away at the
+	// problem rather than outright failing or timing out
+	batchSize := int(batchSizeDeleteSectors)
+	if build.VersionCmp(hostVersion, "1.6.0") < 0 {
+		batchSize = 100
+	}
+
+	// split the indices into batches
+	var batches [][]uint64
+	for {
+		if len(indices) == 0 {
+			break
+		}
+		if len(indices) < batchSize {
+			batchSize = len(indices)
+		}
+		batches = append(batches, indices[:batchSize])
+		indices = indices[batchSize:]
+	}
+
+	// delete the roots
+	var deleted int64
+	err := w.withTransportV2(ctx, hostKey, hostIP, func(t *rhpv2.Transport) error {
```

Comment:
To perform the whole operation atomically I would create a version of `DeleteContractRoots` and `FetchContractRoots` which accept a v2 transport and then have a `PruneContractRoots` for something that locks the contract, runs both of the other method on the same transport and unlocks it again.